org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java

```diff
@@ -150,6 +154,10 @@ protected void triggerValidation(ICompilationUnit cu, long delay) throws JavaMod
 		}
 	}
 
+	private long getDocumentLifecycleDelay() {
+		return Math.min(400, Math.round(1.5 * movingAverage.value));
```

Comment:
some thoughts: 
- for single source of truth, make "400" a constant member of `MovingAverage` (like `initValue` or `maxValue` depending on how you want to use this class)

Reply:
How about:

Keep the `MovingAverage` as a general class without any magic numbers. And extract `400` here as a const and pass it to the constructor of `MovingAverage`. (I'm considering 400ms might not be suitable for other scenario which needs adaptive debounce)

Reply:
Yes, exactly. something like `constructor(public initValue)`, so you can access the initial number to do math here. (init value might not always be a "max" value in other use cases)

Reply:
Sounds good